gasometer/src/lib.rs

```diff
@@ -782,7 +782,9 @@ impl<'config> Inner<'config> {
 				target_exists,
 				..
 			} => costs::suicide_cost(value, target_is_cold, target_exists, self.config),
-			GasCost::SStore { .. } if self.config.estimate => self.config.gas_sstore_set,
+			GasCost::SStore { .. } if self.config.estimate => {
+				self.config.gas_sstore_set + self.config.gas_sload_cold
```

Comment:
Indeed, same fix as https://github.com/rust-blockchain/evm/pull/88

However, having this separate code path for the `GasCost::SStore`, when in estimate mode, seems dangerous and is the reason that led to the estimate to be wrong in the first place (as the EIP-2929 commit (https://github.com/rust-blockchain/evm/commit/b471064011810787960df22194888d5927c6060d) only updated the `sstore_cost` code path).

To avoid similar issues in the future, I would suggest removing this separate path and move the logic into the `sstore_cost` function (as is done in the referenced PR).

Reply:
I have a question, why not just delete the `self.config.estimate` here? It feels like the data will be more accurate

Reply:
I would assume it's because when you are estimating, the state can be different from when you are actually executing the transaction (at a later time). So you want an upper limit when estimating, which is the cost of `gas_sstore_set`.

E.g. I think it could happen that `gas_sstore_reset` would be charged when estimating (based on the state at the time), but when executing `gas_sstore_set` would be charged. Always using `gas_sstore_set` avoids this problem.

Reply:
The EIP-2929 cold transaction costs don't depend on state though, so the `self.config.gas_sload_cold` can be added only if `target_is_cold=true` even if in `estimate` mode.